cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc

```diff
@@ -817,5 +932,126 @@ TEST(TestBinaryArithmetic, AddWithImplicitCastsUint64EdgeCase) {
                                      ArrayFromJSON(uint64(), "[18446744073709551615]")}));
 }
 
+TEST(TestUnaryArithmetic, DispatchBest) {
+  for (std::string name : {"negate"}) {
+    for (const auto& ty : {int8(), int16(), int32(), int64(), uint8(), uint16(), uint32(),
+                           uint64(), float32(), float64()}) {
+      CheckDispatchBest(name, {ty}, {ty});
+      CheckDispatchBest(name, {dictionary(int8(), ty)}, {ty});
+    }
+  }
+
```

Comment:
Please add a test asserting the behavior when `NullType` is used. I expect this will emit a "no kernel found" error (acceptable), or leave the type unpromoted:
```suggestion
  CheckDispatchBest("negate", {null()}, {null()});
```

This would also be fine, but I'd like to see a trivial test case asserting we don't try to access absent buffers:
```c++
  CheckScalarUnary("negate", NullArray(10), NullArray(10));
```

Reply:
Using a `NullType` does trigger a `NotImplemented` error as expected, but I can't catch/assert the error from `CheckDispatchBest()` to pass the tests. Similar scenario for `CheckScalarUnary`. Any ideas?

Reply:
You could add a `CheckDispatchFails(function_name, argument_types)` assert function to go with `CheckDispatchBest()`

Reply:
I added `CheckDispatchFails()` which checks if status is an error after invoking either `DispatchBest()` or `DispatchExact`. It seems the only error produced by `Dispatch*()` is `NotImplemented`. Nevertheless, I made the function description general enough and check for not OK status instead of only `NotImplemented` error, https://github.com/edponce/arrow/blob/ARROW-11950-Compute-Add-unary-negative-kernel/cpp/src/arrow/compute/kernels/test_util.h#L151-L152

Reply:
I did not add `CheckScalarUnary("negate", std::make_shared<NullArray>(10), std::make_shared<NullArray>(10))` test because it fails with `NotImplemented` kernel via a different path of functions. Specifically, `CheckScalar()` --> `CallFunction()` which is where the error is captured. I think adding a family of `CheckScalar*Fails()` function would be beneficial for testing but consider it outside the scope of this PR. Should I open an issue for extending these tests?